### PR TITLE
Fix missing dependecy definition of 'packaging'

### PR DESCRIPTION
**Pull request by s-weigand**
Hi there, we just wanted to upgrade `xarray` to version `0.21.0` and our [ASV workflow on the CI started failing](https://github.com/glotaran/pyglotaran/runs/4989968632?check_suite_focus=true).

```python
File "/home/runner/work/pyglotaran/pyglotaran/benchmark/.asv/env/76391772e92136ec87b9940d70226329/lib/python3.8/site-packages/xarray/core/dask_array_compat.py", line 4, in <module>
       from packaging.version import Version
   ModuleNotFoundError: No module named 'packaging'
```
The reason is that `packaging` isn't a direct dependency of  `xarray`.

```console
$ johnnydep xarray
2022-01-29 08:28:38 [info     ] init johnnydist                [johnnydep.lib] dist=xarray parent=None
2022-01-29 08:28:40 [info     ] init johnnydist                [johnnydep.lib] dist=numpy>=1.18 parent=xarray
2022-01-29 08:28:44 [info     ] init johnnydist                [johnnydep.lib] dist=pandas>=1.1 parent=xarray
2022-01-29 08:28:47 [info     ] init johnnydist                [johnnydep.lib] dist=numpy>=1.18.5 parent=pandas>=1.1
2022-01-29 08:28:51 [info     ] init johnnydist                [johnnydep.lib] dist=python-dateutil>=2.8.1 parent=pandas>=1.1
2022-01-29 08:28:53 [info     ] init johnnydist                [johnnydep.lib] dist=pytz>=2020.1 parent=pandas>=1.1
2022-01-29 08:28:56 [info     ] init johnnydist                [johnnydep.lib] dist=six>=1.5 parent=python-dateutil>=2.8.1
name                            summary
------------------------------  -----------------------------------------------------------------------
xarray                          N-D labeled arrays and datasets in Python
├── numpy>=1.18                 NumPy is the fundamental package for array computing with Python.
└── pandas>=1.1                 Powerful data structures for data analysis, time series, and statistics
    ├── numpy>=1.18.5           NumPy is the fundamental package for array computing with Python.
    ├── python-dateutil>=2.8.1  Extensions to the standard Python datetime module
    │   └── six>=1.5            Python 2 and 3 compatibility utilities
    └── pytz>=2020.1            World timezone definitions, modern and historical
```

The problem why your tests didn't catch this is that `packaging` is a dependency of `pytest` itself.
```console
$ johnnydep pytest
2022-01-29 08:25:03 [info     ] init johnnydist                [johnnydep.lib] dist=pytest parent=None
2022-01-29 08:25:07 [info     ] init johnnydist                [johnnydep.lib] dist=atomicwrites>=1.0 parent=pytest
2022-01-29 08:25:09 [info     ] init johnnydist                [johnnydep.lib] dist=attrs>=19.2.0 parent=pytest
2022-01-29 08:25:11 [info     ] init johnnydist                [johnnydep.lib] dist=colorama parent=pytest
2022-01-29 08:25:14 [info     ] init johnnydist                [johnnydep.lib] dist=iniconfig parent=pytest
2022-01-29 08:25:16 [info     ] init johnnydist                [johnnydep.lib] dist=packaging parent=pytest
2022-01-29 08:25:18 [info     ] init johnnydist                [johnnydep.lib] dist=pluggy<2.0,>=0.12 parent=pytest
2022-01-29 08:25:20 [info     ] init johnnydist                [johnnydep.lib] dist=py>=1.8.2 parent=pytest
2022-01-29 08:25:23 [info     ] init johnnydist                [johnnydep.lib] dist=toml parent=pytest
2022-01-29 08:25:25 [info     ] init johnnydist                [johnnydep.lib] dist=pyparsing!=3.0.5,>=2.0.2 parent=packaging
name                              summary
--------------------------------  ---------------------------------------------------------------------
pytest                            pytest: simple powerful testing with Python
├── atomicwrites>=1.0             Atomic file writes.
├── attrs>=19.2.0                 Classes Without Boilerplate
├── colorama                      Cross-platform colored terminal text.
├── iniconfig                     iniconfig: brain-dead simple config-ini parsing
├── packaging                     Core utilities for Python packages
│   └── pyparsing!=3.0.5,>=2.0.2  Python parsing module
├── pluggy<2.0,>=0.12             plugin and hook calling mechanisms for python
├── py>=1.8.2                     library with cross-python path, ini-parsing, io, code, log facilities
└── toml                          Python Library for Tom's Obvious, Minimal Language
```

We also only saw our ASV workflow fail since ASV only installs the package and all direct runtime dependencies in the venv it runs the benchmark with.

I prepared a [dummy PR](https://github.com/s-weigand/pyglotaran/pull/84) on my fork to demonstrate that this [fixes the issue](https://github.com/s-weigand/pyglotaran/runs/4990111708?check_suite_focus=true).

Since the release is only 10h old maybe just do a quick [post-release](https://www.python.org/dev/peps/pep-0440/#post-releases) so most users won't even notice.

- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

**Comment by s-weigand** [comment]
The failing tests are unrelated to this change and originate from [this commit](https://github.com/pydata/xarray/commit/5470d933452d88deb17cc9294a164c4a03f55dec).

**Comment by mathause** [comment]
Thanks. I think it makes sense to revert #6200 then it's not critical. I opened #6208 for this.

Thanks for the report and the PR.

**Comment by mathause** [comment]
cc @pydata/xarray

**Comment by mathause** [comment]
On second thought I think the problem is the release of dask 2022.01.1.

**Comment by s-weigand** [comment]
> On second thought I think the problem is the release of dask 2022.01.1.
<details>

<summary>
We don't have dask as a dependency
</summary>

```console
$ johnnydep pyglotaran
2022-01-29 10:55:08 [info     ] init johnnydist                [johnnydep.lib] dist=pyglotaran parent=None
2022-01-29 10:55:11 [info     ] init johnnydist                [johnnydep.lib] dist=asteval>=0.9.21 parent=pyglotaran
2022-01-29 10:55:24 [info     ] init johnnydist                [johnnydep.lib] dist=click>=7.0 parent=pyglotaran
2022-01-29 10:55:26 [info     ] init johnnydist                [johnnydep.lib] dist=netCDF4>=1.5.3 parent=pyglotaran
2022-01-29 10:55:29 [info     ] init johnnydist                [johnnydep.lib] dist=numba>=0.52 parent=pyglotaran
2022-01-29 10:55:32 [info     ] init johnnydist                [johnnydep.lib] dist=numpy<1.21,>=1.20.0 parent=pyglotaran
2022-01-29 10:55:36 [info     ] init johnnydist                [johnnydep.lib] dist=pandas>=0.25.2 parent=pyglotaran
2022-01-29 10:55:39 [info     ] init johnnydist                [johnnydep.lib] dist=rich>=10.9.0 parent=pyglotaran
2022-01-29 10:55:42 [info     ] init johnnydist                [johnnydep.lib] dist=ruamel.yaml>=0.17.17 parent=pyglotaran
2022-01-29 10:55:45 [info     ] init johnnydist                [johnnydep.lib] dist=scipy>=1.3.2 parent=pyglotaran
2022-01-29 10:55:49 [info     ] init johnnydist                [johnnydep.lib] dist=sdtfile>=2020.8.3 parent=pyglotaran
2022-01-29 10:55:51 [info     ] init johnnydist                [johnnydep.lib] dist=setuptools>=41.2 parent=pyglotaran
2022-01-29 10:55:54 [info     ] init johnnydist                [johnnydep.lib] dist=tabulate>=0.8.8 parent=pyglotaran
2022-01-29 10:55:56 [info     ] init johnnydist                [johnnydep.lib] dist=xarray!=0.20.0,!=0.20.1,>=0.16.2 parent=pyglotaran
2022-01-29 10:55:58 [info     ] init johnnydist                [johnnydep.lib] dist=colorama parent=click>=7.0
2022-01-29 10:56:00 [info     ] init johnnydist                [johnnydep.lib] dist=cftime parent=netCDF4>=1.5.3
2022-01-29 10:56:03 [info     ] init johnnydist                [johnnydep.lib] dist=numpy>=1.9 parent=netCDF4>=1.5.3
2022-01-29 10:56:07 [info     ] init johnnydist                [johnnydep.lib] dist=numpy>1.13.3 parent=cftime
2022-01-29 10:56:11 [info     ] init johnnydist                [johnnydep.lib] dist=llvmlite<0.39,>=0.38.0rc1 parent=numba>=0.52
2022-01-29 10:56:15 [info     ] init johnnydist                [johnnydep.lib] dist=numpy<1.22,>=1.18 parent=numba>=0.52
2022-01-29 10:56:18 [info     ] init johnnydist                [johnnydep.lib] dist=setuptools parent=numba>=0.52
2022-01-29 10:56:21 [info     ] init johnnydist                [johnnydep.lib] dist=numpy>=1.18.5 parent=pandas>=0.25.2
2022-01-29 10:56:25 [info     ] init johnnydist                [johnnydep.lib] dist=python-dateutil>=2.8.1 parent=pandas>=0.25.2
2022-01-29 10:56:27 [info     ] init johnnydist                [johnnydep.lib] dist=pytz>=2020.1 parent=pandas>=0.25.2
2022-01-29 10:56:29 [info     ] init johnnydist                [johnnydep.lib] dist=six>=1.5 parent=python-dateutil>=2.8.1
2022-01-29 10:56:31 [info     ] init johnnydist                [johnnydep.lib] dist=colorama<0.5.0,>=0.4.0 parent=rich>=10.9.0
2022-01-29 10:56:34 [info     ] init johnnydist                [johnnydep.lib] dist=commonmark<0.10.0,>=0.9.0 parent=rich>=10.9.0
2022-01-29 10:56:36 [info     ] init johnnydist                [johnnydep.lib] dist=pygments<3.0.0,>=2.6.0 parent=rich>=10.9.0
2022-01-29 10:56:38 [info     ] init johnnydist                [johnnydep.lib] dist=ruamel.yaml.clib>=0.2.6 parent=ruamel.yaml>=0.17.17
2022-01-29 10:56:40 [info     ] init johnnydist                [johnnydep.lib] dist=numpy<1.23.0,>=1.16.5 parent=scipy>=1.3.2
2022-01-29 10:56:44 [info     ] init johnnydist                [johnnydep.lib] dist=numpy>=1.15.1 parent=sdtfile>=2020.8.3
2022-01-29 10:56:48 [info     ] init johnnydist                [johnnydep.lib] dist=numpy>=1.18 parent=xarray!=0.20.0,!=0.20.1,>=0.16.2
2022-01-29 10:56:51 [info     ] init johnnydist                [johnnydep.lib] dist=pandas>=1.1 parent=xarray!=0.20.0,!=0.20.1,>=0.16.2
2022-01-29 10:56:55 [info     ] init johnnydist                [johnnydep.lib] dist=numpy>=1.18.5 parent=pandas>=1.1
2022-01-29 10:56:55 [info     ] init johnnydist                [johnnydep.lib] dist=python-dateutil>=2.8.1 parent=pandas>=1.1
2022-01-29 10:56:55 [info     ] init johnnydist                [johnnydep.lib] dist=pytz>=2020.1 parent=pandas>=1.1
2022-01-29 10:56:55 [info     ] init johnnydist                [johnnydep.lib] dist=six>=1.5 parent=python-dateutil>=2.8.1
name                                  summary
------------------------------------  ----------------------------------------------------------------------------------------------------------------

------------
pyglotaran                            The Glotaran fitting engine.
├── asteval>=0.9.21                   Safe, minimalistic evaluator of python expression using ast module
├── click>=7.0                        Composable command line interface toolkit
│   └── colorama                      Cross-platform colored terminal text.
├── netCDF4>=1.5.3                    Provides an object-oriented python interface to the netCDF version 4 library.
│   ├── cftime                        Time-handling functionality from netcdf4-python
│   │   └── numpy>1.13.3              NumPy is the fundamental package for array computing with Python.
│   └── numpy>=1.9                    NumPy is the fundamental package for array computing with Python.
├── numba>=0.52                       compiling Python code using LLVM
│   ├── llvmlite<0.39,>=0.38.0rc1     lightweight wrapper around basic LLVM functionality
│   ├── numpy<1.22,>=1.18             NumPy is the fundamental package for array computing with Python.
│   └── setuptools                    Easily download, build, install, upgrade, and uninstall Python packages
├── numpy<1.21,>=1.20.0               NumPy is the fundamental package for array computing with Python.
├── pandas>=0.25.2                    Powerful data structures for data analysis, time series, and statistics
│   ├── numpy>=1.18.5                 NumPy is the fundamental package for array computing with Python.
│   ├── python-dateutil>=2.8.1        Extensions to the standard Python datetime module
│   │   └── six>=1.5                  Python 2 and 3 compatibility utilities
│   └── pytz>=2020.1                  World timezone definitions, modern and historical
├── rich>=10.9.0                      Render rich text, tables, progress bars, syntax highlighting, markdown and more to the terminal
│   ├── colorama<0.5.0,>=0.4.0        Cross-platform colored terminal text.
│   ├── commonmark<0.10.0,>=0.9.0     Python parser for the CommonMark Markdown spec
│   └── pygments<3.0.0,>=2.6.0        Pygments is a syntax highlighting package written in Python.
├── ruamel.yaml>=0.17.17              ruamel.yaml is a YAML parser/emitter that supports roundtrip preservation of comments, seq/map flow style, and m

ap key order
│   └── ruamel.yaml.clib>=0.2.6       C version of reader, parser and emitter for ruamel.yaml derived from libyaml
├── scipy>=1.3.2                      SciPy: Scientific Library for Python
│   └── numpy<1.23.0,>=1.16.5         NumPy is the fundamental package for array computing with Python.
├── sdtfile>=2020.8.3                 Read Becker & Hickl SDT files
│   └── numpy>=1.15.1                 NumPy is the fundamental package for array computing with Python.
├── setuptools>=41.2                  Easily download, build, install, upgrade, and uninstall Python packages
├── tabulate>=0.8.8                   Pretty-print tabular data
└── xarray!=0.20.0,!=0.20.1,>=0.16.2  N-D labeled arrays and datasets in Python
    ├── numpy>=1.18                   NumPy is the fundamental package for array computing with Python.
    └── pandas>=1.1                   Powerful data structures for data analysis, time series, and statistics
        ├── numpy>=1.18.5             NumPy is the fundamental package for array computing with Python.
        ├── python-dateutil>=2.8.1    Extensions to the standard Python datetime module
        │   └── six>=1.5              Python 2 and 3 compatibility utilities
        └── pytz>=2020.1              World timezone definitions, modern and historical
```
</details>

As far as I can see `xarray/__init__.py` triggers the following import chain :
`xarray.testing`->`xarray.core.duck_array_ops`->`xarray.core.dask_array_compat`  which triggers the import of `packaging`

<details>

<summary>
Traceback starting from our package
</summary>
```python
File "/home/runner/work/pyglotaran/pyglotaran/benchmark/.asv/env/76391772e92136ec87b9940d70226329/lib/python3.8/site-packages/glotaran/io/prepare_dataset.py", line 6, in <module>
       import xarray as xr
     File "/home/runner/work/pyglotaran/pyglotaran/benchmark/.asv/env/76391772e92136ec87b9940d70226329/lib/python3.8/site-packages/xarray/__init__.py", line 1, in <module>
       from . import testing, tutorial, ufuncs
     File "/home/runner/work/pyglotaran/pyglotaran/benchmark/.asv/env/76391772e92136ec87b9940d70226329/lib/python3.8/site-packages/xarray/testing.py", line 8, in <module>
       from xarray.core import duck_array_ops, formatting, utils
     File "/home/runner/work/pyglotaran/pyglotaran/benchmark/.asv/env/76391772e92136ec87b9940d70226329/lib/python3.8/site-packages/xarray/core/duck_array_ops.py", line 24, in <module>
       from . import dask_array_compat, dask_array_ops, dtypes, npcompat, nputils
     File "/home/runner/work/pyglotaran/pyglotaran/benchmark/.asv/env/76391772e92136ec87b9940d70226329/lib/python3.8/site-packages/xarray/core/dask_array_compat.py", line 4, in <module>
       from packaging.version import Version
   ModuleNotFoundError: No module named 'packaging'
```
</details>

**Comment by s-weigand** [comment]
`packaging` is imported in 16 other python files e.g. [indexing.py](https://github.com/pydata/xarray/blob/5470d933452d88deb17cc9294a164c4a03f55dec/xarray/core/indexing.py#L11), I think there is no way around adding it as a dependency. 
Hiding all the imports in closures will be quite error prawn especially since tests won't catch it since `pytest` bring `packaging` along with it.

**Comment by Illviljan** [comment]
https://docs.python.org/3/whatsnew/3.10.html#distutils-deprecated recommends importing `packaging` instead, so if we want to do version checking maybe we should just add it as a dependency?

**Comment by mathause** [comment]
Yes, I think I just expressed it wrong. What I wanted to say: zhe failure you see here is not caused by #6200 but by a new release of dask. And this is independent of your issue with packaging.

**Comment by s-weigand** [comment]
> Yes, I think I just expressed it wrong. What I wanted to say: zhe failure you see here is not caused by #6200 but by a new release of dask. And this is independent of your issue with packaging.

Ahhh now I get it 😅

**Comment by max-sixty** [comment]
Thanks everyone! Are we ok to close this then?

**Comment by keewis** [comment]
no, I think we should still add `packaging` to the dependencies, even if it doesn't cause any failures: we unconditionally import it in at least `xarray.core.pycompat`. We need to add a reasonable minimum version, though.

**Comment by s-weigand** [comment]
Dask has [`packaging>=20`](https://github.com/dask/dask/blob/04ade86291f0f8e7eb6ddd36e06dfb722252345b/setup.py#L35) which was released [over 2y ago](https://pypi.org/project/packaging/#history) guess this is kinda reasonable

**Comment by Illviljan** [comment]
Following dask seems like a good idea! We probably should add packaging in our ci installs as well:
https://github.com/pydata/xarray/tree/main/ci/requirements
and the asv config:
https://github.com/pydata/xarray/blob/main/asv_bench/asv.conf.json

File: setup.cfg [hunk 1]
```diff
@@ -77,6 +77,7 @@ python_requires = >=3.8
 install_requires =
     numpy >= 1.18
     pandas >= 1.1
+    packaging
```
Comment (max-sixty) [inline setup.cfg:80]:
```suggestion
    packaging >= 20
```

**Comment by keewis** [comment]
just for the record, we decided that packaging related tools (i.e. `setuptools`, not sure if that also applies to `packaging`) should be released just before the earliest supported python version, so we would have to decide between `19.2` and `20`. However, since the release notes of `packaging` state that `20` is the first version to officially support `py38` I'd say using `packaging>=20` should be uncontroversial.

**Comment by keewis** [comment]
other than the environment files (in [`ci/requirements`](https://github.com/pydata/xarray/tree/main/ci/requirements)) we also need to update [`doc/getting-started/installing.rst`](https://github.com/pydata/xarray/blob/main/doc/getting-started-guide/installing.rst): the `Required dependencies` section, and maybe also the `Minimum dependency versions`. For the latter, we need to figure out a rule, though.

**Comment by jhamman** [comment]
@pydata/xarray - we should try to get this merged and a new release up ASAP as bug reports are starting to pile up. I've pushed changes to the CI configs, documentation, and elsewhere. I've also pulled in the recent changes on `main` which should hopefully fix the failing CI in the first commit.

**Comment by max-sixty** [comment]
I can do the release later today if no one gets there first

File: ci/install-upstream-wheels.sh [hunk 2]
```diff
@@ -40,6 +40,7 @@ python -m pip install \
     git+https://github.com/zarr-developers/zarr \
     git+https://github.com/Unidata/cftime \
     git+https://github.com/mapbox/rasterio \
+    git+https://github.com/pypa/packaging \
```
Comment (mathause) [inline ci/install-upstream-wheels.sh:43]: Does this to be uninstalled first?

File: doc/contributing.rst [hunk 3]
```diff
@@ -274,13 +274,13 @@ Some other important things to know about the docs:
       .. ipython:: python
 
           x = 2
-          x ** 3
+          x**3
```
Comment (keewis) [inline doc/contributing.rst]:
```suggestion
          x ** 3
```

File: doc/contributing.rst [hunk 4]
```diff
@@ -274,13 +274,13 @@ Some other important things to know about the docs:
...
 
   will be rendered as::
 
       In [1]: x = 2
 
-      In [2]: x ** 3
+      In [2]: x**3
```
Comment (keewis) [inline doc/contributing.rst]:
```suggestion
      In [2]: x ** 3
```

File: doc/whats-new.rst [hunk 5]
```diff
@@ -33,7 +33,8 @@ Deprecations
 
 Bug fixes
 ~~~~~~~~~
-
+- Add `packaging` as a dependency to Xarray (:issue:`6216`, :pull:`6207`).
+  By `Joseph Nowak <https://github.com/josephnowak>`_ and `Joe Hamman <https://github.com/jhamman>`_.
```
Comment (keewis) [inline doc/whats-new.rst:37]: Is that right? Shouldn't it be
```suggestion
  By `Sebastian Weigand <https://github.com/s-weigand>`_ and `Joe Hamman <https://github.com/jhamman>`_.
```

**Review by keewis — COMMENTED** [review]
a few more suggestions, and I think we should also update `requirements.txt`, but after that we should be ready to merge!

**Comment by jhamman** [comment]
This is ready to merge. @max-sixty - whenever you are ready, go ahead and merge and make the release.

**Comment by max-sixty** [comment]
I'll merge now though do the release this eve

**Pull request merged**